provider/pdns.go

```diff
@@ -0,0 +1,369 @@
+					Records:    records,
+					Changetype: string(changetype),
+				}
+
+				// DELETEs explicitly forbid a TTL, therefore only PATCHes need the TTL
+				if changetype == PdnsReplace {
+					if int64(ep.RecordTTL) > int64(math.MaxInt32) {
+						return nil, errors.New("Value of record TTL overflows, limited to int32")
+					}
+					if ep.RecordTTL == 0 {
+						// No TTL was sepecified for the record, we use the default
+						rrset.Ttl = int32(defaultTTL)
+					} else {
+						rrset.Ttl = int32(ep.RecordTTL)
+					}
+				}
+
+				zone.Rrsets = append(zone.Rrsets, rrset)
+
+				// "pop" endpoint if it's matched
+				endpoints = append(endpoints[0:i], endpoints[i+1:]...)
+			} else {
+				// If we didn't pop anything, we move to the next item in the list
+				i++
+			}
+
+		}
+
+		if len(zone.Rrsets) > 0 {
+			zonelist = append(zonelist, zone)
+		}
+
+	}
+
+	// If we still have some endpoints left, it means we couldn't find a matching zone for them
+	// We warn instead of hard fail here because we don't want a misconfig to cause everything to go down
+	if len(endpoints) > 0 {
+		log.Warnf("No matching zones were found for the following endpoints: %+v", endpoints)
+	}
+
+	log.Debugf("Zone List generated from Endpoints: %+v", zonelist)
+
+	return zonelist, nil
+}
+
+// mutateRecords takes a list of endpoints and creates, replaces or deletes them based on the changetype
+func (p *PDNSProvider) mutateRecords(endpoints []*endpoint.Endpoint, changetype pdnsChangeType) error {
+	zonelist, err := p.ConvertEndpointsToZones(endpoints, changetype)
+	if err != nil {
+		return err
+	}
+	for _, zone := range zonelist {
+		jso, err := json.Marshal(zone)
+		if err != nil {
+			log.Debugf("JSON Marshal for zone struct failed!")
+		} else {
+			log.Debugf("Struct for PatchZone:\n%s", string(jso))
+		}
+
+		resp, err := p.client.PatchZone(zone.Id, zone)
+		if err != nil {
+			log.Debugf("PDNS API response: %s", stringifyHTTPResponseBody(resp))
+			return err
+		}
+
+	}
+	return nil
+}
+
+// Records returns all DNS records controlled by the configured PDNS server (for all zones)
+func (p *PDNSProvider) Records() (endpoints []*endpoint.Endpoint, _ error) {
+
+	zones, _, err := p.client.ListZones()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, zone := range zones {
+		z, _, err := p.client.ListZone(zone.Id)
+		if err != nil {
+			log.Warnf("Unable to fetch Records")
+			return nil, err
+		}
+
+		for _, rr := range z.Rrsets {
+			e, err := p.convertRRSetToEndpoints(rr)
+			if err != nil {
+				return nil, err
+			}
+			endpoints = append(endpoints, e...)
+		}
+	}
+
+	log.Debugf("Records fetched:\n%+v", endpoints)
+	return endpoints, nil
+}
+
+// ApplyChanges takes a list of changes (endpoints) and updates the PDNS server
+// by sending the correct HTTP PATCH requests to a matching zone
+func (p *PDNSProvider) ApplyChanges(changes *plan.Changes) error {
+
+	startTime := time.Now()
+
+	// Create
+	for _, change := range changes.Create {
```

Comment:
I would check if debug is true before, otherwise it will loop for nothing everytime the method got called if you don't debug it. Same for Update and Delete, @ffledgling thoughts?

Reply:
Good catch! I'll update code accordingly.

Reply:
changes.Create could also implement Stringer interface `String() string` and then plan.Changes has to know about debug yes/no.

Reply:
Ignoring this change for now because short-circuiting on debug will require a change in the PDNS provider creation struct, which will require a sizable refactor of the code and tests, which after discussion on IRC with @szuecs and @njuettner is more trouble than it's worth at the moment.